topi/python/topi/cuda/deformable_conv2d.py

```diff
@@ -0,0 +1,126 @@
+# pylint: disable=invalid-name
+"""Schedule template of deformable conv2d with cuda backend"""
+import tvm
+from tvm import autotvm
+from .. import nn, generic
+from ..util import traverse_inline
+
+
+autotvm.register_topi_compute(nn.deformable_conv2d_nchw, ["cuda", "gpu"], "direct",
+                              nn.deformable_conv2d_nchw.fdefault)
+
+
+@autotvm.register_topi_schedule(generic.schedule_deformable_conv2d_nchw, ["cuda", "gpu"], "direct")
+def schedule_deformable_conv2d_nchw_cuda(cfg, outs):
+    """TOPI schedule callback of deformable conv2d for cuda gpu
+
+    Parameters
+    ----------
+    cfg: ConfigEntity
+        The config for this template
+
+    outs: Array of Tensor
+        The computation graph description of conv2d
+        in the format of an array of tensors.
+
+    Returns
+    -------
+    s: Schedule
+        The computation schedule for conv2d.
+    """
+    outs = [outs] if isinstance(outs, tvm.tensor.Tensor) else outs
+    s = tvm.create_schedule([x.op for x in outs])
+
+    def _callback(op):
+        if op.tag == 'deformable_conv2d_nchw':
+            schedule_direct_cuda(cfg, s, op.output(0))
+
+    traverse_inline(s, outs[0].op, _callback)
+    return s
+
+
+def schedule_direct_cuda(cfg, s, conv):
+    """Schedule template of deformable conv2d"""
+    n, f, y, x = s[conv].op.axis
+    rc, ry, rx = s[conv].op.reduce_axis
+    cfg.define_split("tile_f", f, num_outputs=4)
+    cfg.define_split("tile_y", y, num_outputs=4)
+    cfg.define_split("tile_x", x, num_outputs=4)
+    cfg.define_split("tile_rc", rc, num_outputs=2)
+    cfg.define_split("tile_ry", ry, num_outputs=2)
+    cfg.define_split("tile_rx", rx, num_outputs=2)
+    cfg.define_knob("auto_unroll_max_step", [0, 512, 1500])
+
+    target = tvm.target.current_target()
+    if target.target_name in ['nvptx', 'rocm']:
+        cfg.define_knob("unroll_explicit", [1])
+    else:
+        cfg.define_knob("unroll_explicit", [0, 1])
+
+    data_deform, kernel = s[conv].op.input_tensors
+
+    s[data_deform].compute_inline()
+    if isinstance(kernel.op, tvm.tensor.ComputeOp) and 'dilate' in kernel.op.tag:
+        s[kernel].compute_inline()
+
+    if conv.op in s.outputs:
+        output = conv
+        OL = s.cache_write(conv, 'local')
+    else:
+        output = s.outputs[0].output(0)
+        s[conv].set_scope('local')
+        OL = conv
+
+    # create cache stage
+    AA = s.cache_read(data_deform, 'shared', [OL])
+    WW = s.cache_read(kernel, 'shared', [OL])
+
```

Comment:
just comment, if the deformable group value is small, maybe loading pixel offsets into shared memory may help boost performance (so offsets loading can be shared across channels)

Reply:
Thanks, I will try it